constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+2. User clicks a point. A variable `p1` is defined with the screen coordinates.
+3. User clicks another point. A variable `p2` is defined at that point. Since they're in line-sketch mode, `l1 = line(p1, p2)` is also generated. The app knows to name it with the prefix `l` because it is generating the line, so it knows it's a line. Similarly with points.
+4. Repeat the above for points `p3` and `p4`.
+5. When they close the sketch by clicking `p1`, the app recognizes this, does not define a fifth point, and creates `l4 = line(p4, p1)`, closing the sketch.
+6. The user selects one of the lines, clicks "Constrain length with function". This opens an input box in the command palette where the user can type an arbitrary expression.
+7. The user types `0.5 * ||`, and the UI knows that the user wants to select a piece of geometry.
+8. The user selects a line with their pointing device, and the app fills in `l1||`.
+9. The user types `^2 Enter`, which finishes the expression.
+10. At this point, the app appends at the end of the code the new line `||l3|| = 0.5 * ||l1||^2`.
+	- Design note: We could add this relationship next to the definition of `l3`. This is a reasonable choice. But I think we should prefer to append to the end by default because it retains the history of user operations in the order that they did them, which can be useful when reading code. Not only do UI operations correspond one-to-one with lines of code (a spatial relationship), but they correspond in order to how the author created them (a temporal relationship). This strengthens the tie between the UI and code in the user's mind. If they find that they'd like a different order, they can reorder lines without affecting the meaning of the code in many cases. (We'll discuss exactly which cases later.)
+11. Assume similar for `||l2||`.
+
+## First-Class Points
+
+Why name points instead of embedding the coordinates in the `line()` function calls? Mechanical engineers who are used to other CAD software expect to be able to create a point as standalone geometry. In this proposal, points are first-class. This has a few benefits:
+
+1. It fits ME expectations better
+2. It allows creating points that are used purely for construction (not display or manufacturing). For example, users can create other geometry derived from the point through relationships.
+3. Named points are natural snap points in the point-and-click UI. Similar to how we use variable names to autocomplete in the editor, when the user is drawing, we can use named points as the set of points that a user might want to snap to, which is the point-and-click analogue of autocomplete.
+4. It allows creating unconstrained points.
+	1. In the above example, we used `p1 = screen_point(x: 10, y: 12)`, which is fully constrained to a coordinate relative to the screen's coordinate system.
+	2. However, we could make an unconstrained point `p1 = point()`. This is still useful for constraining in code later.
```

Comment:
I could see `screen_point` being seen as the "default" way of making points confusing people writing KCL (not who I think we should prioritize), and `point` seeming more like an abstract power user way of defining a point. These seem powerful, but I would make the concrete one the shortest and easiest to write.